### DISCUSSION TITLE
Strictly define what the terms "parsing", "serializing", "deserializing", and "type conversion" mean and how they relate.

### DISCUSSION THREAD

**Discussion by gregmarr** [nlohmann/json#3540] [Ideas]
@falbrechtskirchinger wrote:
> Some day we should strictly define what the terms "parsing", "serializing", "deserializing", and "type conversion" mean and how they relate.

My thoughts:

-  parsing/loading: conversion of text/binary blob/file/stream to `nlohmann::basic_json`
-  deserializing: conversion of `nlohmann::basic_json` to `ArbitraryStructure`
-  serializing: conversion of `ArbitraryStructure` to `nlohmann::basic_json`
-  dumping/outputting/saving: conversion of `nlohmann::basic_json` to  text/binary blob/file/stream.

**Comment by nlohmann** [comment]
My 2 cents:

- "Parsing" and "deserializing" are synonyms for any function converting a serialization (i.e., sequence of bytes) into a structured format. API functions are `parse` or `from_xxx` for the binary formats.
- "Serializing" and "dumping", etc. are synonyms for the reverse function of converting a structure to a sequence of bytes. API functions are `dump` or `to_xxx` for the binary formats.
- The conversions with `to_json` and `from_json` have no proper name to me. In the end they can be seen as casting between C++ types.

That said, it's good to have a discussion! I guess this will greatly improve the documentation.

**Comment by falbrechtskirchinger** [comment]
When I made that comment I had a somewhat different idea. Having read two definitions, I don't know anymore. I'd defer to the dictionary and compare how other well-known projects use the terms.
